### PR TITLE
Add an analyzer plugin.

### PR DESCRIPTION
**Pull request by davidmorgan**
Major parts to this change;

 - Add a plugin using the plugin API. This part was already reviewed by the analyzer team.
 - Refactor built_value_generator so that when it produces an error and suggested fix it also gives the source location. This is then enough for the plugin to use. Basically, it takes the errors and suggested fixes that you'd get at generation time and offers to apply them for you.
 - Tests for the plugin that check what it does to various code samples.

I consider the plugin alpha quality at this point, so it's not yet referred to in the docs.

---
This change is [Reviewable](https://reviewable.io/reviews/google/built_value.dart/382)

### CODE REVIEW

File: analysis_options.yaml [hunk 1]
```diff
@@ -1,7 +1,19 @@
 analyzer:
   strong-mode:
     implicit-casts: false
-#    implicit-dynamic: false
+    # implicit-dynamic: false
+
+  # Add this to your project `analysis_options.yaml` to enable the built_value
+  # analyzer plugin. You should also add the top level `plugins` clause below.
```
Comment (manuelF) [inline analysis_options.yaml]: Maybe a 5-word summary on what the plugin does here would be good for people copy-pasting it.

File: built_value_generator/lib/src/plugin/checker.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        if (!type.interfaces.any((i) => i.displayName.startsWith('Built')))
+          continue;
+
+        final ValueSourceClass sourceClass = new ValueSourceClass(type);
+        final errors = sourceClass.computeErrors();
+
+        if (errors.isNotEmpty) {
+          final lineInfo = compilationUnit.lineInfo;
+
+          // Report one error on the 'Built' interface. Bundle together all the
```
Comment (manuelF) [inline built_value_generator/lib/src/plugin/checker.dart]: line too long

File: built_value_generator/lib/src/plugin/checker.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                  compilationUnit.source.fullName,
+                  offset,
+                  length,
+                  offsetLineLocation.lineNumber,
+                  offsetLineLocation.columnNumber),
+              'Class needs fixes for built_value: ' +
+                  errors.map((error) => error.message).join(' '),
+              'BUILT_VALUE_NEEDS_FIXES');
+
+          // Fix consists of all the individual fixes, sorted so they apply from
```
Comment (manuelF) [inline built_value_generator/lib/src/plugin/checker.dart]: line too long

File: built_value_generator/lib/src/plugin/plugin.dart [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  void processResult(AnalysisResult analysisResult) {
+    try {
+      if (analysisResult.unit == null ||
+          analysisResult.libraryElement == null) {
+        channel.sendNotification(
+            new plugin.AnalysisErrorsParams(analysisResult.path, [])
+                .toNotification());
+        return;
+      }
+      final checkResult = checker.check(analysisResult.libraryElement);
```
Comment (manuelF) [inline built_value_generator/lib/src/plugin/plugin.dart]: Add a comment why this is a separate case then the previous one, it is not obvious to me what additional info you get by sending extra data.

File: built_value_generator/lib/src/value_source_class.dart [hunk 5]
```diff
@@ -39,7 +40,14 @@ abstract class ValueSourceClass
       name.startsWith('_') ? '_\$${name.substring(1)}' : '_\$$name';
 
   @memoized
-  ClassElement get builderElement => element.library.getType(name + 'Builder');
+  ClassElement get builderElement {
+    final result = element.library.getType(name + 'Builder');
+    if (result == null) return null;
+    // If the builder is in a generated file, then we're analyzing _after_ code
```
Comment (manuelF) [inline built_value_generator/lib/src/value_source_class.dart]: line too long

File: built_value_generator/lib/src/value_source_class.dart [hunk 6]
```diff
@@ -258,110 +264,213 @@ abstract class ValueSourceClass
...
+        ..fix = 'abstract '));
     }
 
     if (hasBuiltValueImportWithShow) {
-      result.add('Stop using "show" when importing '
-          '"package:built_value/built_value.dart". It prevents the generated '
-          'code from finding helper methods.');
+      result.add(new GeneratorError((b) => b
+        ..message = 'Stop using "show" when importing '
+            '"package:built_value/built_value.dart". It prevents the generated '
```
Comment (manuelF) [inline built_value_generator/lib/src/value_source_class.dart]: line too long

File: analysis_options.yaml (hunk 1 above)
Comment (davidmorgan) [reply analysis_options.yaml]: Done.

File: built_value_generator/lib/src/plugin/checker.dart (hunk 2 above)
Comment (davidmorgan) [reply built_value_generator/lib/src/plugin/checker.dart]: This one's 80 cols which I believe is okay.

File: built_value_generator/lib/src/plugin/checker.dart (hunk 3 above)
Comment (davidmorgan) [reply built_value_generator/lib/src/plugin/checker.dart]: Done.

File: built_value_generator/lib/src/plugin/plugin.dart (hunk 4 above)
Comment (davidmorgan) [reply built_value_generator/lib/src/plugin/plugin.dart]: Done.

File: built_value_generator/lib/src/value_source_class.dart (hunk 5 above)
Comment (davidmorgan) [reply built_value_generator/lib/src/value_source_class.dart]: This also looks like 80 cols--should be okay.

File: built_value_generator/lib/src/value_source_class.dart (hunk 6 above)
Comment (davidmorgan) [reply built_value_generator/lib/src/value_source_class.dart]: Done.

**Pull request merged**